### ISSUE TITLE
Change how scripts are handled

### ISSUE THREAD

**Issue by salahchafai** [stride3d/stride#803]
**Is your feature request related to a problem? Please describe.**
Currently there are 3 different types of scripts (startup, sync and async), i think it'll be better if startup and sync scripts get unified (and maybe async too) and have more functions like FixedUpdate, Awake, LateUpdate, OnEnable, OnDisable (virtual so you can override only the ones you need), the current system isn't scalabe as new methods require new classes (which is not ideal), so having a single class (maybe next to async) that has all the functions and then the user can override the ones he needs would be super helpful, also using virtual methods, or interfaces is standard with a lot of component based frameworks. As such, this is how most users would expect it to function. finally i think it'll be way better to have scripts activable so you can easily enable and disable scripts (which is very useful)

**Describe the solution you'd like**
Have a single class (let's say ScriptBehavior or something) that contains all the virtual functions, and make the scriptComponent inherit from ActivableEntityComponent instead of EntityComponent

**Describe alternatives you've considered**
The user can add these functions by himself (in async script for example) but that's not a great way of doing things and it's unnecessary extra work imo

**Comment by tebjan** [comment]
Yes having a few more convenient methods to override would be nice.

This could be done in a new script type that takes care of that logic and provides these as virtual methods to override by the user.

I'd not remove the previous script types, that would make all existing projects invalid...

**Comment by salahchafai** [comment]
@tebjan actually no, we can have some kind of a class responsible for upgrading the scripts so if it's a sync script it gets changed to the new type, if it's a startup script it gets changed to the new type and only override the start function, also i'll most likely keep the async script so that will work with no extra work), having many script types only causes confusion imo

**Comment by Kryptos-FR** [comment]
> FixedUpdate, Awake, LateUpdate, OnEnable, OnDisable

Looks like you want to mimic Unity. But Stride doesn't work like Unity.

**Comment by salahchafai** [comment]
@Kryptos-FR i don't want to mimic Unity, many people expect these functions (and many other engines have them like Unigine), and these are just functions i suggested and are not necessarily (the ones i see important are FixedUpdate, OnEnable, OnDisable)

**Comment by manio143** [comment]
Would you consider providing those methods in the async context first?
While I understand your motivation it seems to me that creating a ScriptBehaviour with optionally overridable methods and using reflection to check those would be bad architecture.
On the other hand, extending the ScriptSystem to support `await Script.FixedUpdate()` would achieve the goal and get more people to use the async features of Stride.
The user could create their own type with overrideable methods should they want that:
```csharp
public class OverrideableScript : StartupScript {
    public virtual void Update() {}
    public virtual void FixedUpdate() {}
    public override void Start() {
        Script.AddTask(_Update);
        Script.AddTask(_FixedUpdate);
    }
    public async Task _Update() { while(true) { await Script.NextFrame(); Update(); }}
    public async Task _FixedUpdate() { while(true) { await Script.FixedUpdate(); FixedUpdate(); }}
}
```

**Comment by salahchafai** [comment]
@manio143 await is slow so it's not ideal for having a large amount of scripts, also the check will only happen when the script is added (for the first time) and it's not heavy, and not all functions are gonna be checked (Init(), OnEnable(), OnDisable() will just run without having to be checked since they'll run directly (without getting registred to the scheduler)

**Comment by manio143** [comment]
I did some tests and yes, the async code has a larger overhead than the sync code. In general though, the scheduler adds overhead also to sync methods.

Anyway, I believe this feature will require quite a few modifications around the scheduler.

So, when creating the mechanisms for a sync scenario, it would be a good idea to expose the mentioned methods for async scripts as well.
I also agree with tebjan about leaving existing types. In particular they may be subclasses of the ScriptBehaviour.

**Comment by salahchafai** [comment]
some functions like FixedUpdate and OnDraw might be addable to the async script i agree (through await), however others like Init aren't really a loop or something so they'll only be available in sync, i'm still not sure about the other types tbh, i'll for sure keep the async but are startup and sync really useful ? you'll be able to use their functionality through the new class and i'll make a class responsible for upgrading older projects automatically (from startup and sync to the new type)
